File: tools/unpack_exepack.py
#!/usr/bin/env python3
"""
Strip the Microsoft EXEPACK layer from build/KBU1.EXE and emit a flat
build/KBU2.EXE with a proper DOS relocation table.

Pipeline for the whole game (everything after the pristine game/ originals
lands in build/):
    game/KB.EXE  --(tools/unpack_nwc.py)-->  build/KBU1.EXE
                 --(this script)---------->  build/KBU2.EXE

Run with no arguments to use those default paths; pass src/dst to override.

KB.EXE is double-packed: an outer custom New World Computing packer (stripped
by unpack_nwc.py) wrapping an inner Microsoft EXEPACK layer (removed here).
KBU2.EXE is the flat, uncompressed translation base: every string sits at a
fixed offset, editable in place.

Validated: KBU2's decompressed image matches the live running game (dump1)
98.9% after relocation; the ~1% delta is runtime-mutated variables.

The EXEPACK format
------------------
The packer stub is identified by the signature "RB" at its CS:0x10, and by its
error string "Packed file is corrupt". Its 18-byte header at CS:0000 holds the
real entry CS:IP, SS:SP and the decompressed length.

Decompression reads records BACKWARD from stub CS:0000, expanding the image
upward in place. Any trailing 0xFF padding is skipped first, then each record is
    [cmd][len_hi][len_lo]
with `cmd & 0xFE` selecting the operation -- 0xB0 = fill (one fill byte follows
below the record), 0xB2 = copy a literal run -- and `cmd & 1` marking the last
record.

EXEPACK self-relocates at runtime from its own table, which is why KBU1.EXE
runs at any load address despite declaring 0 relocations in its DOS header. That
table sits right after the "Packed file is corrupt" string as 16 sections, each
a uint16 count followed by that many uint16 offsets; section i contributes
segment base i*0x1000. It is decoded here and written out as a real 1960-entry
DOS relocation table, which is what makes KBU2.EXE an ordinary loadable EXE.
"""
import os, struct, sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from paths import KBU1, KBU2                                 # noqa: E402


def unpack(src_path=KBU1, dst_path=KBU2,
           validate=None, validate_base=0x8920, load_seg=0x892):
    d = open(src_path, "rb").read()
    # Take the load image exactly as the DOS header declares it -- KBU1's
    # header is 32 bytes and its image 107501 bytes, but reading the fields
    # keeps this honest if the file is ever regenerated differently. (CUP386's
    # output carried ~78 KB of dump tail past the declared end; the declared
    # length is what DOS would load, and what the EXEPACK offsets below assume.)
    cblp, cp, _crlc, cparhdr = struct.unpack_from("<4H", d, 2)
    img = bytearray(d[cparhdr * 16:(cp - 1) * 512 + cblp])

    # EXEPACK header lives at the packer stub's CS:0000. For KBU1 that's image
    # offset 0x19350 (stub entry CS:IP = 1935:0012; IP 0x12 = past the 18-byte hdr).
    cs0 = 0x19350
    (real_ip, real_cs, _mem_start, exepack_size,
     real_sp, real_ss, dest_len, skip_len) = struct.unpack_from("<8H", img, cs0)
    assert bytes(img[cs0 + 16:cs0 + 18]) == b"RB", "EXEPACK 'RB' signature not found"

    # ---- decompress EXEPACK RLE, reading records backward from cs0 ----
    dest = bytearray(dest_len * 16)
    dest[:cs0] = img[:cs0]              # packed bytes live low, expand in place upward
    src = cs0
    while src > 0 and img[src - 1] == 0xFF:   # skip 0xFF padding before first record
        src -= 1
    dst = len(dest)
    records = 0
    while True:
        cmd = img[src - 1]; length = img[src - 3] | (img[src - 2] << 8); src -= 3
        op = cmd & 0xFE
        if op == 0xB0:                 # fill: next byte repeated `length` times
            fill = img[src - 1]; src -= 1
            dest[dst - length:dst] = bytes([fill]) * length; dst -= length
        elif op == 0xB2:               # copy: `length` literal bytes
            dest[dst - length:dst] = dest[src - length:src]; src -= length; dst -= length
        else:
            raise ValueError(f"bad EXEPACK opcode 0x{cmd:02x}")
        records += 1
        if cmd & 1:                    # low bit set = last record
            break

    # ---- decode EXEPACK's own relocation table (16 sections after the error string) ----
    p = img.find(b"Packed file is corrupt") + 22
    relocs = []
    for section in range(16):
        cnt = struct.unpack_from("<H", img, p)[0]; p += 2
        for _ in range(cnt):
            off = struct.unpack_from("<H", img, p)[0]; p += 2
            relocs.append((section * 0x1000, off))

    out = bytes(dest)

    # ---- optional byte-level validation against a live memory dump ----
    if validate:
        try:
            gt = open(validate, "rb").read()[validate_base:validate_base + dest_len * 16]
            applied = bytearray(out)
            for seg, off in relocs:
                lin = seg * 16 + off
                w = (applied[lin] | (applied[lin + 1] << 8)) + load_seg & 0xffff
                applied[lin] = w & 0xff; applied[lin + 1] = w >> 8
            same = sum(1 for a, b in zip(applied, gt) if a == b)
            print(f"validation vs {validate}: {100 * same / len(gt):.2f}% match after relocation")
        except FileNotFoundError:
            pass

    # ---- write flat MZ with a real relocation table ----
    reloc_bytes = b"".join(struct.pack("<HH", off, seg) for seg, off in relocs)
    hdr_size = ((0x1C + len(reloc_bytes) + 511) // 512) * 512
    total = hdr_size + len(out)
    minalloc = max(0x200, (real_ss * 16 + real_sp - len(out)) // 16 + 0x200)
    hdr = bytearray(hdr_size)
    struct.pack_into("<14H", hdr, 0, 0x5A4D, total % 512, (total + 511) // 512,
                     len(relocs), hdr_size // 16, minalloc, 0xFFFF,
                     real_ss, real_sp, 0, real_ip, real_cs, 0x1C, 0)
    hdr[0x1C:0x1C + len(reloc_bytes)] = reloc_bytes
    open(dst_path, "wb").write(bytes(hdr) + out)
    print(f"wrote {dst_path}: {total} bytes, {len(relocs)} relocations, "
          f"{records} EXEPACK records, entry {real_cs:04x}:{real_ip:04x}, SS:SP {real_ss:04x}:{real_sp:04x}")

if __name__ == "__main__":
    unpack(*(sys.argv[1:3] or []))

File: core/unpackExepack.ts
/**
 * Strips EXEPACK, rebuilding the relocation table: the result is the edit base.
 * Flat and uncompressed, every string at a fixed offset, editable in place.
 *
 * The packer stub is identified by the signature "RB" at its CS:0x10, and by its
 * error string "Packed file is corrupt". Its 18-byte header at CS:0000 holds the
 * real entry CS:IP, SS:SP and the decompressed length.
 *
 * Decompression reads records BACKWARD from stub CS:0000, expanding the image
 * upward in place. Any trailing 0xFF padding is skipped first, then each record
 * is
 *     [cmd][len_hi][len_lo]
 * with `cmd & 0xFE` selecting the operation -- 0xB0 = fill (one fill byte
 * follows below the record), 0xB2 = copy a literal run -- and `cmd & 1` marking
 * the last record.
 *
 * EXEPACK self-relocates at runtime from its own table, which is why the input
 * runs at any load address despite declaring 0 relocations in its DOS header.
 * That table sits right after the "Packed file is corrupt" string as 16
 * sections, each a uint16 count followed by that many uint16 offsets; section i
 * contributes segment base i*0x1000. It is decoded here and written out as a
 * real DOS relocation table, which is what makes the output an ordinary loadable
 * EXE.
 */

import { concat, equal, find, fromLatin1, setU16s, u16, u16s } from "./bytes.ts";

/**
 * Image offset of the packer stub's CS:0000, where the EXEPACK header sits: the
 * stub's entry is CS:IP = 1935:0012, IP 0x12 being just past that 18-byte header.
 */
const STUB = 0x19350;

const CORRUPT_MSG = "Packed file is corrupt";

export const unpackExepack = (packed: Uint8Array): Uint8Array => {
  // The load image exactly as the DOS header declares it, never the file size:
  // the declared length is what DOS would load, and what the EXEPACK offsets
  // below assume. Any tail past it is not part of the image.
  const [cblp, cp, , cparhdr] = u16s(packed, 2, 4);
  const img = packed.subarray(cparhdr * 16, (cp - 1) * 512 + cblp);

  const [realIp, realCs, , , realSp, realSs, destLen] = u16s(img, STUB, 8);
  if (!equal(img.subarray(STUB + 16, STUB + 18), fromLatin1("RB"))) {
    throw new Error(`EXEPACK 'RB' signature not found at 0x${(STUB + 16).toString(16)}`);
  }

  // Packed bytes live low and expand in place upward.
  const dest = new Uint8Array(destLen * 16);
  dest.set(img.subarray(0, STUB));
  let src = STUB;
  while (src > 0 && img[src - 1] === 0xff) src--; // 0xFF padding before the first record
  let dst = dest.length;
  for (;;) {
    const cmd = img[src - 1];
    const length = img[src - 3] | (img[src - 2] << 8);
    src -= 3;
    if ((cmd & 0xfe) === 0xb0) {
      const fill = img[src - 1];
      src -= 1;
      dest.fill(fill, dst - length, dst);
      dst -= length;
    } else if ((cmd & 0xfe) === 0xb2) {
      dest.copyWithin(dst - length, src - length, src);
      src -= length;
      dst -= length;
    } else {
      throw new Error(`bad EXEPACK opcode 0x${cmd.toString(16).padStart(2, "0")}`);
    }
    if (cmd & 1) break; // low bit set = last record
  }

  const msg = find(img, fromLatin1(CORRUPT_MSG));
  if (msg === -1) throw new Error(`"${CORRUPT_MSG}" not found; no EXEPACK relocation table`);
  let p = msg + CORRUPT_MSG.length;
  const relocs: { seg: number; off: number }[] = [];
  for (let section = 0; section < 16; section++) {
    const count = u16(img, p);
    p += 2;
    for (let i = 0; i < count; i++) {
      relocs.push({ seg: section * 0x1000, off: u16(img, p) });
      p += 2;
    }
  }

  const relocBytes = new Uint8Array(relocs.length * 4);
  relocs.forEach((r, i) => setU16s(relocBytes, i * 4, [r.off, r.seg]));
  const hdrSize = Math.ceil((0x1c + relocBytes.length) / 512) * 512;
  const total = hdrSize + dest.length;
  const minalloc = Math.max(0x200, Math.floor((realSs * 16 + realSp - dest.length) / 16) + 0x200);
  const hdr = new Uint8Array(hdrSize);
  // prettier-ignore
  setU16s(hdr, 0, [0x5a4d, total % 512, Math.ceil(total / 512), relocs.length, hdrSize / 16,
                   minalloc, 0xffff, realSs, realSp, 0, realIp, realCs, 0x1c, 0]);
  hdr.set(relocBytes, 0x1c);
  return concat([hdr, dest]);
};
